iree/base/tracing.h

```diff
@@ -12,93 +12,424 @@
+
+#endif  // IREE_TRACING_FEATURES
+
+#ifdef __cplusplus
+}  // extern "C"
+#endif  // __cplusplus
+
+//===----------------------------------------------------------------------===//
+// Instrumentation macros (C)
+//===----------------------------------------------------------------------===//
+
+// Matches Tracy's PlotFormatType enum.
+enum {
+  // Values will be displayed as plain numbers.
+  IREE_TRACING_PLOT_TYPE_NUMBER = 0,
+  // Treats the values as memory sizes. Will display kilobytes, megabytes, etc.
+  IREE_TRACING_PLOT_TYPE_MEMORY = 1,
+  // Values will be displayed as percentage with value 100 being equal to 100%.
+  IREE_TRACING_PLOT_TYPE_PERCENTAGE = 2,
+};
+
+// Colors used for messages based on the level provided to the macro.
+enum {
+  IREE_TRACING_MESSAGE_LEVEL_ERROR = 0xFF0000u,
+  IREE_TRACING_MESSAGE_LEVEL_WARNING = 0xFFFF00u,
+  IREE_TRACING_MESSAGE_LEVEL_INFO = 0xFFFFFFu,
+  IREE_TRACING_MESSAGE_LEVEL_VERBOSE = 0xC0C0C0u,
+  IREE_TRACING_MESSAGE_LEVEL_DEBUG = 0x00FF00u,
+};
+
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
 
-#if defined(WTF_ENABLE)
+// Sets an application-specific payload that will be stored in the trace.
+// This can be used to fingerprint traces to particular versions and denote
+// compilation options or configuration. The given string value will be copied.
+#define IREE_TRACE_SET_APP_INFO(value, value_length) \
+  ___tracy_emit_message_appinfo(value, value_length)
 
-#include "wtf/event.h"   // IWYU pragma: export
-#include "wtf/macros.h"  // IWYU pragma: export
+// Sets the current thread name to the given string value.
+// This will only set the thread name as it appears in the tracing backend and
+// not set the OS thread name as it would appear in a debugger.
+// The C-string |name| will be copied and does not need to be a literal.
+#define IREE_TRACE_SET_THREAD_NAME(name) iree_tracing_set_thread_name_impl(name)
 
-namespace iree {
+// Begins a new zone with the parent function name.
+#define IREE_TRACE_ZONE_BEGIN(zone_id) \
+  IREE_TRACE_ZONE_BEGIN_NAMED(zone_id, NULL)
 
-// Initializes tracing if it is built into the binary.
-// Does nothing if already initialized.
-void InitializeTracing();
+// Begins a new zone with the given compile-time literal name.
+#define IREE_TRACE_ZONE_BEGIN_NAMED(zone_id, name_literal)                    \
+  static const struct ___tracy_source_location_data TracyConcat(              \
+      __tracy_source_location, __LINE__) = {name_literal, __FUNCTION__,       \
+                                            __FILE__, (uint32_t)__LINE__, 0}; \
+  iree_zone_id_t zone_id = iree_tracing_zone_begin_impl(                      \
+      &TracyConcat(__tracy_source_location, __LINE__), NULL, 0);
 
-// Returns whether tracing support is compiled into the binary.
-bool IsTracingAvailable();
+// Begins a new zone with the given runtime dynamic string name.
+// The |value| string will be copied into the trace buffer.
```

Comment:
nit: each of these functions has a different way of saying "the string will be copied" :P

> // The |value| string will be copied into the trace buffer.

> // The |file_name|, |function_name|, and optional |name| strings will be copied
> // into the trace buffer and do not need to persist.

> // The |value| string will be copied and can be freed after calling.

(no action needed)